### ISSUE TITLE
Custom Monitor noted as in-use and cannot be removed

### ISSUE THREAD

**Issue by paulpaulini** [F5Networks/f5-appsvcs-extension#110]
##### Do you already have an issue opened with F5 support?
YES - C3018563

##### ISSUE TYPE

 - Bug Report

##### AS3 BUILD/ VERSION
f5-appsvcs-3.11.0-3

##### BIGIP VERSION
13.1.1.4

##### SUMMARY
I have seen an issue with AS3 when using custom monitors in the declaration.

The following declaration uses two monitors for the pool "web_pool", and http type monitor and a custom monitor "test_monitor". 
Posting the declaration works fine, but after reposting the declaration with a changed monitor (e.g. from http to tcp) the declaration can´t be deleted anymore.

The only thing that helps is a "touch /service/mcpd/forceload" and 
reboot, afterwards the declaration can be deleted via AS3 straight away without errors.

##### STEPS TO REPRODUCE
1) Post following declaration:
```json
{
    "class": "AS3",
    "action": "deploy",
    "persist": true,
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.11.0",
        "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
        "label": "Sample 1",
        "remark": "Simple HTTP Service with Round-Robin Load Balancing",
        "updateMode": "selective",
        "Sample_01": {
            "class": "Tenant",
            "A1": {
                "class": "Application",
                "template": "shared",
                "serviceMain": {
                    "class": "Service_HTTP",
                    "virtualAddresses": [
                        "10.0.1.10"
                    ],
                    "pool": "web_pool"
                },
                "web_pool": {
                    "class": "Pool",
                    "monitors": [
                        "http",
                        {
                            "use": "test_monitor"
                        }
                    ],
                    "minimumMonitors": 1,
                    "members": [
                        {
                            "servicePort": 80,
                            "serverAddresses": [
                                "192.0.1.10",
                                "192.0.1.11"
                            ]
                        }
                    ]
                },
                "test_monitor": {
                    "class": "Monitor",
                    "label": "test_monitor",
                    "monitorType": "http",
                    "interval": 30,
                    "timeout": 91,
                    "send": "GET /\r\n",
                    "receive": "GET / HTTP/1.1\r\nHost: www.test.com\r\nConnection: Close\r\n\r\n"
                }
            }
        }
    }
}
```
2) Re-POST the declaration with a changed monitor (from http to tcp):
```json
{
    "class": "AS3",
    "action": "deploy",
    "persist": true,
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.11.0",
        "id": "urn:uuid:33045210-3ab8-4636-9b2a-c98d22ab915d",
        "label": "Sample 1",
        "remark": "Simple HTTP Service with Round-Robin Load Balancing",
        "updateMode": "selective",
        "Sample_01": {
            "class": "Tenant",
            "A1": {
                "class": "Application",
                "template": "shared",
                "serviceMain": {
                    "class": "Service_HTTP",
                    "virtualAddresses": [
                        "10.0.1.10"
                    ],
                    "pool": "web_pool"
                },
                "web_pool": {
                    "class": "Pool",
                    "monitors": [
                        "tcp",
                        {
                            "use": "test_monitor"
                        }
                    ],
                    "minimumMonitors": 1,
                    "members": [
                        {
                            "servicePort": 80,
                            "serverAddresses": [
                                "192.0.1.10",
                                "192.0.1.11"
                            ]
                        }
                    ]
                },
                "test_monitor": {
                    "class": "Monitor",
                    "label": "test_monitor",
                    "monitorType": "http",
                    "interval": 30,
                    "timeout": 91,
                    "send": "GET /\r\n",
                    "receive": "GET / HTTP/1.1\r\nHost: www.test.com\r\nConnection: Close\r\n\r\n"
                }
            }
        }
    }
}
```

3) Try to delete the declaration by sending a DELETE to
https: //{{bigip_a_mgmt}}/mgmt/shared/appsvcs/declare/Sample_01

Following error occurs:
```json
{
    "results": [
        {
            "message": "declaration failed",
            "response": "01070083:3: Monitor /Sample_01/A1/test_monitor is in use.",
            "code": 422,
            "host": "localhost",
            "tenant": "Sample_01",
            "runTime": 3137
        }
    ],
    "declaration": {
        "Sample_01": {
            "class": "Tenant",
            "A1": {
                "class": "Application",
                "template": "shared",
                "serviceMain": {
                    "class": "Service_HTTP",
                    "virtualAddresses": [
                        "10.0.1.10"
                    ],
                    "pool": "web_pool"
                },
                "web_pool": {
                    "class": "Pool",
                    "monitors": [
                        "tcp",
                        {
                            "use": "test_monitor"
                        }
                    ],
                    "minimumMonitors": 1,
                    "members": [
                        {
                            "servicePort": 80,
                            "serverAddresses": [
                                "192.0.1.10",
                                "192.0.1.11"
                            ]
                        }
                    ]
                },
                "test_monitor": {
                    "class": "Monitor",
                    "label": "test_monitor",
                    "monitorType": "http",
                    "interval": 30,
                    "timeout": 91,
                    "send": "GET /\r\n",
                    "receive": "GET / HTTP/1.1\r\nHost: www.test.com\r\nConnection: Close\r\n\r\n"
                }
            }
        },
        "class": "ADC",
        "schemaVersion": "3.0.0",
        "id": "1560414757478",
        "updateMode": "selective",
        "controls": {
            "archiveTimestamp": "2019-06-13T08:32:40.817Z"
        }
    },
    "code": 422
}
```
##### EXPECTED RESULTS

Configuration Change

##### ACTUAL RESULTS

Error
Failure m
essage: "message": "declaration failed",
"response": "01070083:3: Monitor 
/Sample_01/A1/test_monitor is in use.",

**Comment by focrensh** [comment]
I just tested and noticed that the monitor does not change from http to tcp even though it shows success.

Additional Test:

- Send empty array of monitors > success 
- Update declaration to only http or tcp > failure: the test_monitor is also re-applied.

AT-499 opened for tracking the bug.

**Comment by ansmithf5** [comment]
Good afternoon @paulpaulini,

This is actually a known issue with BIG-IP and monitors, not AS3. The bug is due to some checking BIG-IP does with monitors, which makes it appear as though the monitor is still in use.

As a work around, if you post the original declaration again. You should be able to DELETE it. Then after the DELETE, POST the 2nd one again. The 2nd configuration will succeed.

**Comment by notandy** [comment]
Hi any update on this? I have the issue too, just that i cannot delete any custom monitors without removing them from the pools upfront, but it should be possible at the same time.

**Comment by dstokesf5** [comment]
This is a bug with BIG-IP, and is not something that will be resolved by AS3. If you are encountering this issue, it is recommended that you use the provided workaround: Delete the tenant in AS3, and then submit the full new tenant declaration.

**Comment by notandy** [comment]
> This is a bug with BIG-IP, and is not something that will be resolved by AS3. If you are encountering this issue, it is recommended that you use the provided workaround: Delete the tenant in AS3, and then submit the full new tenant declaration.

Is this issue tracked by F5? I am using AS3 for automatic configuration of BigIPs via OpenStack, so deleting a tenant would be no solution and would impose severe outages. I found at least a workaround by just derefencing the healthmonitors before deleting them. But I would really like to see this fixed, do you have a contact or ticket that I can follow up? I could try to engage some traction to get this fixed in BigIP.

**Comment by kwinkelman-bw** [comment]
Second the question about internal tracking/bug report - I am running into this as well, and also do not find the solution of deleting the entire tenant to be workable in our environment.

**Comment by wncocz** [comment]
I've been able to work around this (in our cases, at least) by changing pools to "monitor none" prior to AS3 re-deploy.  It seems like AS3 could do something like this on the backend after a BIG-IP version check.  As ugly as it is, it would be better than having to do it manually on our end.  F5 supports both pieces of software; it seems like teams could work together.

**Comment by dstokesf5** [comment]
It looks like the internal bz is 527004.

**Comment by dstokesf5** [comment]
Re-opening this in our internal product backlog as AUTOTOOL-1469 to take another stab at working around the tmsh bug.

**Comment by jasonkradams** [comment]
This is resolved with Application Services Release v3.21.0.

**Comment by walterk82** [comment]
> I have a support case open and I would suggest that you open one as well to inquire about BZ ID 747020

I'm sorry BZ ID 747020 is not related to this issue.

**Comment by jasonkradams** [comment]
> > I have a support case open and I would suggest that you open one as well to inquire about BZ ID 747020
> 
> I'm sorry BZ ID 747020 is not related to this issue.

Apologies; mixed up my BZIDs. v3.21.0 has implemented a workaround for ID 527004 [Bug ID 527004: Monitor delete-and-create within a transaction fails](https://cdn.f5.com/product/bugtracker/ID527004.html)

**Comment by walterk82** [comment]
There is also a new BZID 

807337 -- Config  utility (web UI) output differs between tmsh and AS3 when the pool monitor is   changed.

I'm not sure how they are releated.

**Comment by jasonkradams** [comment]
> There is also a new BZID
> 
> 807337 -- Config utility (web UI) output differs between tmsh and AS3 when the pool monitor is changed.
> 
> I'm not sure how they are releated.

Both have to do with 'cli transactions' which is what AS3 does to actually push the changes. AS3 will take the declaration then build a `cli script`, then execute that script inside of a `cli transaction`. Both of those IDs have to do with failure conditions inside of a `cli transaction`.

Here is the public information for [Bug ID 807337: Config utility (web UI) output differs between tmsh and AS3 when the pool monitor is changed.](https://cdn.f5.com/product/bugtracker/ID807337.html)

I have been informed that  ID 807337 will be fixed in the next incremental release of BIG-IP (14.1.2.7, 15.1.0.5, etc). I am not sure if the AS3 team has implemented a workaround for BIG-IP Bug ID 807337 or not.

**Comment by dstokesf5** [comment]
There has been no AS3 work-around implemented for BZ807337.

**Comment by dstokesf5** [comment]
This has been resolved as part of the 3.21.0 release.
